### PR TITLE
Require Thread Master Key to be 16 bytes.

### LINKED ISSUE #1700: otThreadSetMasterkey
The api call is given by:
```
ThreadError otThreadSetMasterKey(otInstance *aInstance, const uint8_t *aKey, uint8_t aKeyLength)
```
If this is called with a masterkey that is not 16 bytes in length, e.g.
```
uint8_t key = {0x5E, 0xA1, 0xED};
otThreadSetMasterKey(aInstance, const &key, 3);
```
The 16 byte masterkey buffer will look like
```
0x5E, 0xA1, 0xED, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88, 0x99, 0xAA, 0xBB, 0xCC, 0xDD, 0xEE, 0xFF
```
because of the default masterkey that was used before. (One could consider clearing the buffer first)

The above results in the following
1. The actual masterkey used is: `0x5E, 0xA1, 0xED, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00`
2. Getting the masterkey results in `0x5E, 0xA1, 0xED, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88, 0x99, 0xAA, 0xBB, 0xCC, 0xDD, 0xEE, 0xFF` with length 16

If another thread devices joins, it will get the entire 16 byte buffer as masterkey. Since Thread does not provide a mechanic to communicate the masterkey's length to the new device, the joiner router and the joiner will have a different master key.

There are two issues:
1. The length field does not seem to be taken into account consistently (code problem)
2. The length field cannot be communicated to a new joiner device (api flaw)

### PR DESCRIPTION
**Pull request by jwhui**
- Update code to use `otMasterKey` type.

Resolves #1700.

### CODE REVIEW

File: examples/drivers/windows/otApi/otApi.cpp [hunk 1]
```diff
@@ -1598,51 +1598,37 @@ otThreadSetLinkMode(
...
     {
         free(Result);
         return nullptr;
     }
-    else
-    {
-        *aKeyLength = Result->Length;
-    }
-    return (uint8_t*)Result;
+    return (otMasterKey*)Result;
```
Comment (nibanks) [inline examples/drivers/windows/otApi/otApi.cpp]: Shouldn't need to cast here, right?

File: examples/drivers/windows/otApi/otApi.cpp [hunk 2]
```diff
@@ -1598,51 +1598,37 @@ otThreadSetLinkMode(
...
     if (aInstance == nullptr) return kThreadError_InvalidArgs;
     
-    BYTE Buffer[sizeof(GUID) + sizeof(otMasterKey) + sizeof(uint8_t)];
+    BYTE Buffer[sizeof(GUID) + sizeof(otMasterKey)];
     memcpy_s(Buffer, sizeof(Buffer), &aInstance->InterfaceGuid, sizeof(GUID));
-    memcpy_s(Buffer + sizeof(GUID), sizeof(Buffer) - sizeof(GUID), aKey, aKeyLength);
-    memcpy_s(Buffer + sizeof(GUID) + sizeof(otMasterKey), sizeof(Buffer) - sizeof(GUID) - sizeof(otMasterKey), &aKeyLength, sizeof(aKeyLength));
+    memcpy_s(Buffer + sizeof(GUID), sizeof(Buffer) - sizeof(GUID), aKey, sizeof(otMasterKey));
     
     return DwordToThreadError(SendIOCTL(aInstance->ApiHandle, IOCTL_OTLWF_OT_MASTER_KEY, Buffer, sizeof(Buffer), nullptr, 0));
```
Comment (nibanks) [inline examples/drivers/windows/otApi/otApi.cpp]: Should be able to do something like this now instead:

```
DwordToThreadError(SetIOCTL(aInstance, IOCTL_OTLWF_OT_MASTER_KEY, aKey));
```

File: examples/apps/windows/otAdapter.h [hunk 3]
```diff
@@ -296,39 +296,32 @@ public ref class otAdapter sealed
...
             }
             else
             {
-                key[keyLen++] = (uint8_t)(charToValue(value->Data()[0]));
-                for (uint32_t i = 1; i < value->Length(); i += 2)
-                {
-                    key[keyLen++] = (uint8_t)((charToValue(value->Data()[i]) << 4) |
-                        charToValue(value->Data()[i + 1]));
-                }
+                throw Exception::CreateException(E_INVALIDARG);
```
Comment (nibanks) [inline examples/apps/windows/otAdapter.h]: I'd prefer to make this code pad the end with zero than require a full key.

File: src/ncp/ncp_base.cpp [hunk 4]
```diff
@@ -2258,17 +2258,16 @@ ThreadError NcpBase::GetPropertyHandler_NET_XPANID(uint8_t header, spinel_prop_k
 ThreadError NcpBase::GetPropertyHandler_NET_MASTER_KEY(uint8_t header, spinel_prop_key_t key)
 {
     const uint8_t *ptr(NULL);
-    uint8_t len(0);
 
-    ptr = otThreadGetMasterKey(mInstance, &len);
+    ptr = otThreadGetMasterKey(mInstance)->m8;
```
Comment (abtink) [inline src/ncp/ncp_base.cpp]: Suggestion: let's get rid of the `ptr` here and we can do this instead:
```
    return SendPropertyUpdate(
                 header,
                 SPINEL_CMD_PROP_VALUE_IS,
                 key,
                 SPINEL_DATATYPE_DATA_S,
                 otThreadGetMasterKey(mInstance)->m8,
                 OT_MASTER_KEY_SIZE
             );
  }
```

**Pull request merged**